python/shotgun_model/data_handler.py

```diff
@@ -0,0 +1,473 @@
+        """
+        Convenience wrapper around warning logging
+
+        :param msg: debug message
+        """
+        self._bundle.log_warning("[%s] %s" % (self.__class__.__name__, msg))
+
+    def _sg_clean_data(self, sg_data):
+        """
+        Recursively clean the supplied SG data for use by clients.
+
+        This method currently handles:
+
+            - Converting datetime objects to universal time stamps.
+
+        :param sg_data: Shotgun data dictionary
+        :return: Cleaned up Shotgun data dictionary
+        """
+        # Older versions of Shotgun return special timezone classes. QT is
+        # struggling to handle these. In fact, on linux it is struggling to
+        # serialize any complex object via QDataStream. So we need to account
+        # for this for older versions of SG.
+        #
+        # Convert time stamps to unix time. Unix time is a number representing
+        # the timestamp in the number of seconds since 1 Jan 1970 in the UTC
+        # timezone. So a unix timestamp is universal across time zones and DST
+        # changes.
+        #
+        # When you are pulling data from the shotgun model and want to convert
+        # this unix timestamp to a *local* timezone object, which is typically
+        # what you want when you are displaying a value on screen, use the
+        # following code:
+        # >>> local_datetime = datetime.fromtimestamp(unix_time)
+        #
+        # furthermore, if you want to turn that into a nicely formatted string:
+        # >>> local_datetime.strftime('%Y-%m-%d %H:%M')
+
+        if isinstance(sg_data, dict):
+            for k in sg_data.keys():
+                sg_data[k] = self._sg_clean_data(sg_data[k])
+        elif isinstance(sg_data, list):
+            for i in range(len(sg_data)):
+                sg_data[i] = self._sg_clean_data(sg_data[i])
+        elif isinstance(sg_data, datetime.datetime):
+            # convert to unix timestamp, local time zone
+            sg_data = time.mktime(sg_data.timetuple())
+
+        return sg_data
+
+    def _sg_compare_data(self, a, b):
+        """
+        Compares two shotgun data structures.
+        Both inputs are assumed to contain utf-8 encoded data.
+
+        :returns: True if a is same as b, false otherwise
+        """
+        if isinstance(a, dict):
+            if not isinstance(b, dict):
+                return False
+            if len(a) != len(b):
+                return False
+            for a_key in a.keys():
+                if not self._sg_compare_data(a.get(a_key), b.get(a_key)):
+                    return False
+
+        elif isinstance(a, list):
+            if not isinstance(b, list):
+                return False
+            if len(a) != len(b):
+                return False
+            for idx in xrange(len(a)):
+                if not self._sg_compare_data(a[idx], b[idx]):
+                    return False
+
+        # handle thumbnail fields as a special case
```

Comment:
haha... i was just looking at the previous conditionals which are totally generic and then *boom* we're in `amazonaws` land. probably fine to leave it for now.